### PR TITLE
New Source: Commcare

### PR DESCRIPTION
**Pull request by siddhant3030**
## What

Commcare is a digital data collection platform for front-line workers working in offline settings. For organizations using Commcare having a native Airbyte connector to import data will allow for ingestion into their data warehouses and use of the data for analysis. Current processes involve ingestion with Python scripts, maintaining schemas ad-hoc in the scripts, and through Excel downloads.

## How

The current version of the connector provides for ingesting data through Airbyte from multiple forms in an organization's Commcare application. A typical use case of a Commcare system involves a root project space in which multiple applications can be hosted and can be defined for multiple program interventions. The current connector defines the ingestion of data at the application level as that is the most common entry point for Commcare Applications. The implemented version of the connector will allow for all sync modes - full refresh and incremental including de-duped mode. A best-guess normalization is also provided to ease the normalization process using a couple of python libraries to guess the schema from the data fetched back through the API.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

Screenshot 2022-12-09 at 12 01 38 AM

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

Screenshot 2023-01-05 at 11 31 50 PM

</details>

### CODE REVIEW

**Comment by geekwhocodes** [comment]
@sajarin I can review this PR.

**Comment by siddhant3030** [comment]
Screenshot 2022-12-12 at 11 58 57 AM

Flake check

**Comment by geekwhocodes** [comment]
Hi @siddhant3030, Thanks for your amazing contribution! 
Could you please delete unused files?
- source-commcare\source_commcare\schemas\TODO.md
- source-commcare\sample_files

Also, can you upload a screenshot of `integration tests`?

It would be good if you move custom classes to separate file/s.

**Comment by sajarin** [comment]
@siddhant3030 can you respond to the requested changes ASAP? Thanks!

**Review by geekwhocodes — CHANGES_REQUESTED** [review]

**Comment by siddhant3030** [comment]
@geekwhocodes Thank you for the suggestions. I'll work on this. Also we are fixing the acceptance test. As soon as we're done we can let you know. I'll also add the screenshot.

**Comment by sajarin** [comment]
@siddhant3030 any update here? We're trying to get this merged before the end of the week. Otherwise, the merging might be significantly delayed due to the holidays. We're fairly close to finishing the review here.

File: airbyte-integrations/connectors/source-commcare/source_commcare/source.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            # Airbyte converts the cursor_field value (datetime) to string when it saves the state and
+            # our state setter parses the saved state with a format that contains microseconds
+            # self._cursor_value must have non-zero microseconds for the formatting and parsing to work correctly.
+            # This issue would also occur if an incoming record had a timestamp with zero microseconds
+            self._cursor_value = self._cursor_value.replace(microsecond=10)
+        # This cycle of pull is complete so clear out the form ids we saved for this cycle
+        CommcareStream.forms.clear()
+
+
+class Form(IncrementalStream):
```
[the remark this answers is not in the document]
Comment (siddhant3030) [reply airbyte-integrations/connectors/source-commcare/source_commcare/source.py:189]: what are doc strings ?

**Comment by siddhant3030** [comment]
@sajarin will try to push this today

**Comment by sajarin** [comment]
@siddhant3030 any update here? Trying to get this merged by EOD today. Let me know if you need help with the requested changes.

**Comment by siddhant3030** [comment]
@sajarin We are working on some config changes and making it more general to use this connector. We will let you know latest by Monday. Thankyou for your replies.

File: airbyte-integrations/connectors/source-commcare/source_commcare/source.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+            # self._cursor_value must have non-zero microseconds for the formatting and parsing to work correctly.
+            # This issue would also occur if an incoming record had a timestamp with zero microseconds
+            self._cursor_value = self._cursor_value.replace(microsecond=10)
+
+
+# Source
+class SourceCommcare(AbstractSource):
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        if "api_key" not in config:
+            # print("Returning No")
```
[the remark this answers is not in the document]
Comment (siddhant3030) [reply airbyte-integrations/connectors/source-commcare/source_commcare/source.py:252]: done

File: airbyte-integrations/connectors/source-commcare/source_commcare/source.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+            "authenticator": auth,
+        }
+        appdata = Application(**{**args, "app_id": config["app_id"]}).read_records(sync_mode=SyncMode.full_refresh)
+
+        # Generate streams for forms, one per xmlns and one stream for cases.
+        streams = self.generate_streams(args, config, appdata)
+        return streams
+
+    def generate_streams(self, args, config, appdata):
+        form_args = {**args, "app_id": config["app_id"], "start_date": config["start_date"]}
```
[the remark this answers is not in the document]
Comment (siddhant3030) [reply airbyte-integrations/connectors/source-commcare/source_commcare/source.py]: done

File: airbyte-integrations/connectors/source-commcare/source_commcare/source.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+                    name2xmlns[name] = xmlns
+
+        # Create the streams from the collected names
+        # Sorted by name
+        for k in sorted(name2xmlns):
+            key = name2xmlns[k]
+            stream = Form(**form_args, name=k, xmlns=key, schema=self.empty_schema())
+            streams.append(stream)
+
+        stream = Case(**args, app_id=config["app_id"], start_date=config["start_date"], schema=self.empty_schema())
```
[the remark this answers is not in the document]
Comment (siddhant3030) [reply airbyte-integrations/connectors/source-commcare/source_commcare/source.py]: done

File: airbyte-integrations/connectors/source-commcare/source_commcare/source.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+                        formname = f"Unnamed_{xmlns[-5:]}"
+
+                    name = formname
+                    name2xmlns[name] = xmlns
+
+        # Create the streams from the collected names
+        # Sorted by name
+        for k in sorted(name2xmlns):
+            key = name2xmlns[k]
+            stream = Form(**form_args, name=k, xmlns=key, schema=self.empty_schema())
```
[the remark this answers is not in the document]
Comment (siddhant3030) [reply airbyte-integrations/connectors/source-commcare/source_commcare/source.py]: done

File: airbyte-integrations/connectors/source-commcare/source_commcare/source.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+                    break
+            if found:
+                self._cursor_value = datetime.strptime(record[self.cursor_field], self.dateformat)
+                record.update({"streamname": "case", "indexed_on": record["indexed_on"] + "Z"})  # Make indexed_on tz aware
+                # convert xform_ids field from array to comma separated list so flattening won't create
+                # one field per item. This is because some cases have up to 2000 xform_ids and we don't want 2000 extra
+                # fields in the schema
+                record["xform_ids"] = ",".join(record["xform_ids"])
+                frec = flatten(record)
+                # print(frec)
```
[the remark this answers is not in the document]
Comment (siddhant3030) [reply airbyte-integrations/connectors/source-commcare/source_commcare/source.py]: done

File: airbyte-integrations/connectors/source-commcare/source_commcare/source.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+
+        params = {"format": "json"}
+        return params
+
+
+class Application(CommcareStream):
```
[the remark this answers is not in the document]
Comment (siddhant3030) [reply airbyte-integrations/connectors/source-commcare/source_commcare/source.py:65]: done

File: airbyte-integrations/connectors/source-commcare/source_commcare/source.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+            params.update(next_page_token)
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        for o in iter(response.json()["objects"]):
+            yield o
+        return None
+
+
+class Case(IncrementalStream):
```
[the remark this answers is not in the document]
Comment (siddhant3030) [reply airbyte-integrations/connectors/source-commcare/source_commcare/source.py:140]: done

**Comment by siddhant3030** [comment]
@geekwhocodes @sajarin can you help me with this? This is the only test case which is failing in acceptance test

Screenshot 2022-12-12 at 11 58 57 AM

**Comment by marcosmarxm** [comment]
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

**Comment by siddhant3030** [comment]
@geekwhocodes @sajarin any updates on this? Needed your help the acceptance this. I have posted the query above

**Review by sajarin — APPROVED** [review]

**Comment by sajarin** [comment]
/publish connector=connectors/source-commcare run-tests=false

**Pull request merged**